package org.ota.service;

import java.util.concurrent.*;

/**
 * 描述：
 *
 * @author liuyangyang
 * @create 2018-09-19 15:57
 **/
public class ThreadPoolExecutorDemo extends ThreadPoolExecutor {
    public static volatile Integer num = 1;
    public ThreadPoolExecutorDemo() {
//        super(3, 5,
//                60,
//                TimeUnit.SECONDS,
//                new LinkedBlockingQueue<Runnable>(3));
//        super(3, 5,
//                60,
//                TimeUnit.SECONDS,
//                new LinkedBlockingQueue<Runnable>(3),new DiscardPolicy());
        super(3, 5,
                60,
                TimeUnit.SECONDS,
                new LinkedBlockingQueue<Runnable>(3),new RejectedExecutionHandler()
                {
                    public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
                        System.out.println("aiai"+Thread.currentThread().getName());
                    }
                });
    }

    public static void main(String[] args) {
        ThreadPoolExecutorDemo demo1=new ThreadPoolExecutorDemo();
        for (int i=1;i<20;i++)
        {
            demo1.execute(new Runnable() {
                @Override
                public void run() {

                    if(num <= 5)
                    {
                        num++;
                        throw  new RuntimeException("123"+Thread.currentThread().getName());
                    }
                    System.out.println("hi");
                    try {
                        Thread.sleep(6000000);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                    System.out.println("***");
                }
            });
        }

        System.out.println(demo1.getActiveCount());
        System.out.println(demo1.getQueue().size());
        demo1.shutdown();
    }
}


Exception in thread "pool-1-thread-1" java.lang.RuntimeException: 123pool-1-thread-1
	at org.ota.service.ThreadPoolExecutorDemo$2.run(ThreadPoolExecutorDemo.java:44)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Exception in thread "pool-1-thread-3" java.lang.RuntimeException: 123pool-1-thread-3
	at org.ota.service.ThreadPoolExecutorDemo$2.run(ThreadPoolExecutorDemo.java:44)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Exception in thread "pool-1-thread-5" java.lang.RuntimeException: 123pool-1-thread-5
	at org.ota.service.ThreadPoolExecutorDemo$2.run(ThreadPoolExecutorDemo.java:44)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Exception in thread "pool-1-thread-2" java.lang.RuntimeException: 123pool-1-thread-2
	at org.ota.service.ThreadPoolExecutorDemo$2.run(ThreadPoolExecutorDemo.java:44)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Exception in thread "pool-1-thread-4" java.lang.RuntimeException: 123pool-1-thread-4
	at org.ota.service.ThreadPoolExecutorDemo$2.run(ThreadPoolExecutorDemo.java:44)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
hi
hi
hi
hi
hi
aiaimain
aiaimain
aiaimain
aiaimain
aiaimain
aiaimain
5
3


package org.ota.service;

import java.util.concurrent.*;

/**
 * 描述：
 *
 * @author liuyangyang
 * @create 2018-09-19 15:57
 **/
public class ThreadPoolExecutorDemo extends ThreadPoolExecutor {
    public static volatile Integer num = 1;
    public ThreadPoolExecutorDemo() {
//        super(3, 5,
//                60,
//                TimeUnit.SECONDS,
//                new LinkedBlockingQueue<Runnable>(3));
//        super(3, 5,
//                60,
//                TimeUnit.SECONDS,
//                new LinkedBlockingQueue<Runnable>(3),new DiscardPolicy());
        super(3, 5,
                60,
                TimeUnit.SECONDS,
                new LinkedBlockingQueue<Runnable>(3),new RejectedExecutionHandler()
                {
                    public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
                        System.out.println("aiai"+Thread.currentThread().getName());
                    }
                });
    }

    public static void main(String[] args) {

        ThreadPoolExecutorDemo demo1=new ThreadPoolExecutorDemo();

        try{
            for (int i=1;i<20;i++)
            {
                demo1.execute(new Runnable() {
                    @Override
                    public void run() {

                        if(num <= 5)
                        {
                            num++;
                            throw  new RuntimeException("123"+Thread.currentThread().getName());
                        }
                        System.out.println("hi");
                        try {
                            Thread.sleep(6000000);
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                        System.out.println("***");
                    }
                });
            }
        }catch (Exception e)
        {
            System.out.println("error"+e);
        }

        System.out.println(demo1.getActiveCount());
        System.out.println(demo1.getQueue().size());

        demo1.shutdown();
    }
}
Exception in thread "pool-1-thread-1" Exception in thread "pool-1-thread-3" Exception in thread "pool-1-thread-4" Exception in thread "pool-1-thread-5" java.lang.RuntimeException: 123pool-1-thread-5
	at org.ota.service.ThreadPoolExecutorDemo$2.run(ThreadPoolExecutorDemo.java:47)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Exception in thread "pool-1-thread-2" java.lang.RuntimeException: 123pool-1-thread-2
	at org.ota.service.ThreadPoolExecutorDemo$2.run(ThreadPoolExecutorDemo.java:47)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
java.lang.RuntimeException: 123pool-1-thread-1
	at org.ota.service.ThreadPoolExecutorDemo$2.run(ThreadPoolExecutorDemo.java:47)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
java.lang.RuntimeException: 123pool-1-thread-3
	at org.ota.service.ThreadPoolExecutorDemo$2.run(ThreadPoolExecutorDemo.java:47)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)hi
hi
aiaimain
hi

	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
java.lang.RuntimeException: 123pool-1-thread-4
	at org.ota.service.ThreadPoolExecutorDemo$2.run(ThreadPoolExecutorDemo.java:47)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
hi
hi
aiaimain
aiaimain
aiaimain
aiaimain
aiaimain
5
3
